sycl/source/detail/device_image_impl.hpp

```diff
@@ -150,6 +185,49 @@ class device_image_impl {
   }
 
 private:
+  void updateSpecConstSymMap() {
+    if (MBinImage) {
+      const pi::DeviceBinaryImage::PropertyRange &SCRange =
+          MBinImage->getSpecConstants();
+      using SCItTy = pi::DeviceBinaryImage::PropertyRange::ConstIterator;
+
+      // This variable is used to calculate spec constant value offset in a
+      // flat byte array.
+      unsigned BlobOffset = 0;
+      for (SCItTy SCIt : SCRange) {
+        const char *SCName = (*SCIt)->Name;
+
+        pi::ByteArray Descriptors =
+            pi::DeviceBinaryProperty(*SCIt).asByteArray();
+        assert(Descriptors.size() > 8 && "Unexpected property size");
+
+        // Expected layout is vector of 3-component tuples (flattened into a
+        // vector of scalars), where each tuple consists of: ID of a scalar spec
+        // constant, (which might be a member of the composite); offset, which
+        // is used to calculate location of scalar member within the composite
+        // or zero for scalar spec constants; size of a spec constant
+        constexpr size_t NumElements = 3;
+        assert(((Descriptors.size() - 8) / sizeof(std::uint32_t)) %
+                       NumElements ==
+                   0 &&
+               "unexpected layout of composite spec const descriptors");
+        auto *It = reinterpret_cast<const std::uint32_t *>(&Descriptors[8]);
+        auto *End = reinterpret_cast<const std::uint32_t *>(&Descriptors[0] +
+                                                            Descriptors.size());
+        while (It != End) {
+          // The map is not locked here because updateSpecConstSymMap() is only
+          // supposed to be called from c'tor.
+          MSpecConstSymMap[std::string{SCName}].push_back(
+              SpecConstDescT{/*ID*/ It[0], /*CompositeOffset*/ It[1],
+                             /*Size*/ It[2], BlobOffset});
+          BlobOffset += /*Size*/ It[2];
```

Comment:
@alexbatashev, sorry for late comment, but this is not quite correct. `BlobOffset` should be calculated using specialization constants sorted by their numeric IDs in ascending order, while in properties they are listed in a random order (because we are using `map` in `sycl-post-link` for them, they are sorted by name instead of numeric ID).

Reply:
Ok, I'll fix that